writer.go

```diff
@@ -0,0 +1,145 @@
+
+	pineconeClient, indexConnection, err := NewPineconeClient(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pinecone pineconeClient: %v", err)
+	}
+
+	writer := &Writer{
+		client: pineconeClient,
+		index:  indexConnection,
+	}
+	return writer, nil
+}
+
+func (w *Writer) Upsert(ctx context.Context, record sdk.Record) error {
+	ID := recordID(record.Key)
+
+	payload, err := recordPayload(record.Payload)
+	if err != nil {
+		return fmt.Errorf("error getting payload: %v", err)
+	}
+
+	metadata, err := recordMetadata(record.Metadata)
+	if err != nil {
+		return fmt.Errorf("error getting metadata: %v", err)
+	}
+
+	sdk.Logger(ctx).Error().Msgf("metadata: %v", metadata)
+
+	vector := []*pinecone.Vector{{
+		Id:     ID,
+		Values: payload,
+		SparseValues: nil,
+		Metadata:     metadata,
+	}}
+
+	_, err = w.index.UpsertVectors(&ctx, vector)
+	if err != nil {
+		return fmt.Errorf("error upserting record: %v ", err)
+	}
+
+	sdk.Logger(ctx).Trace().Msg("upserted record")
+	return nil
+}
+
+// Delete deletes records by a key.
+func (w *Writer) Delete(ctx context.Context, record sdk.Record) error {
+	ID := []string{recordID(record.Key)}
+
+	err := w.index.DeleteVectorsById(&ctx, ID)
+	if err != nil {
+		return fmt.Errorf("\n error deleting record: %v ", err)
+	}
+
+	sdk.Logger(ctx).Trace().Msg("Successful record delete.")
+	return nil
+}
+
+func (w *Writer) Close() error {
+	return w.index.Close()
+}
+
+// NewPineconeClient takes the Pinecone Index URL string in Config and splits into respective parts to establish a
+// connection
+func NewPineconeClient(ctx context.Context, config DestinationConfig) (*pinecone.Client, *pinecone.IndexConnection, error) {
+	sdk.Logger(ctx).Trace().Msg("Creating a Pinecone Client.")
+
+	client, err := pinecone.NewClient(pinecone.NewClientParams{
+		ApiKey: config.PineconeAPIKey,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating Pinecone client: %v", err)
+	}
+	sdk.Logger(ctx).Info().Msg("created pinecone client")
+
+	// index urls should have their protocol trimmed
+	hostURL := strings.TrimPrefix(config.PineconeHostURL, "https://")
+
+	index, err := client.Index(hostURL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error establishing index connection: %v", err)
+	}
+	sdk.Logger(ctx).Info().Msgf("created pinecone index")
+
+	return client, index, nil
+}
+
+func recordID(Key sdk.Data) string {
+	key := Key.Bytes()
+	return string(key)
+}
+
+func recordPayload(payload sdk.Change) ([]float32, error) {
+	data := payload.After
+
+	if data == nil || len(data.Bytes()) == 0 {
+		return nil, errors.New("empty payload")
+	}
+
+	var parsedPayload []float32
+
+	err := json.Unmarshal(data.Bytes(), &parsedPayload)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling JSON: %v", err)
+	}
+
+	return parsedPayload, nil
+}
+
+func recordMetadata(data sdk.Metadata) (*pinecone.Metadata, error) {
```

Comment:
Nitpick: for the `recordXXX` methods here, wdyt about renaming to `getPineconeXXX`? When I read `recordMetadat` it made me think we're getting the OpenCDC record metadata, but it's actually Pinecone metadata. Similary for the payload, we're getting vectors.